File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.java
import java.util.*;

class Solution {
    public int solution(int[][] rectangle, int characterX, int characterY, int itemX, int itemY) {
       final int N = 100;

        int[][] map = new int[N + 1][N + 1];
        boolean[][] visit = new boolean[N + 1][N + 1];

        for (int[] a : rectangle) {
            int x1 = a[0] * 2;
            int y1 = a[1] * 2;
            int x2 = a[2] * 2;
            int y2 = a[3] * 2;

            for (int x = x1; x <= x2; x++) {
                for (int y = y1; y <= y2; y++) {

                    if (x == x1 || x == x2 || y == y1 || y == y2) {
                        if (map[x][y] == 0) {
                            map[x][y] = 1;
                        }
                    }
                    else {
                        map[x][y] = 2;
                    }
                }
            }
        }

        Queue<Point> qu = new ArrayDeque<>();
        qu.offer(new Point(characterX * 2, characterY * 2, 0));

        visit[characterX * 2][characterY * 2] = true;

        int[] dx = {-1, 1, 0, 0};
        int[] dy = {0, 0, -1, 1};

        while (!qu.isEmpty()) {

            Point now = qu.poll();

            int x = now.x;
            int y = now.y;
            int move = now.move;

            if (x == itemX * 2 && y == itemY * 2) {
                return move / 2;
            }

            for (int i = 0; i < 4; i++) {
                int nx = x + dx[i];
                int ny = y + dy[i];

                if (nx < 0 || ny < 0 || nx > N || ny > N || map[nx][ny] != 1 || visit[nx][ny]) {
                    continue;
                }

                visit[nx][ny] = true;
                qu.offer(new Point(nx, ny, move + 1));
            }
        }

        return 0;
    }

    static class Point {

        int x, y, move;

        public Point(int x, int y, int move) {
            this.x = x;
            this.y = y;
            this.move = move;
        }
    }
}

File: 프로그래머스/3/87694. 아이템 줍기/아이템 줍기.py
from collections import deque


def bfs(characterX, characterY, itemX, itemY, N, map):
    visit = [[False] * (N + 1) for _ in range(N + 1)]
    sx, sy = characterX * 2, characterY * 2

    visit[sx][sy] = True
    qu = deque()
    qu.append((sx, sy, 0))

    dx = (-1, 1, 0, 0)
    dy = (0, 0, -1, 1)

    while qu:
        x, y, move = qu.popleft()

        if x == itemX * 2 and y == itemY * 2:
            return move // 2

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx < 0 or ny < 0 or nx > N or ny > N or map[nx][ny] != 1 or visit[nx][ny]:
                continue

            visit[nx][ny] = True
            qu.append((nx, ny, move + 1))


def solution(rectangle, characterX, characterY, itemX, itemY):
    N = 100
    map = [[0] * (N + 1) for _ in range(N + 1)]

    for a in rectangle:
        x1, y1 = a[0] * 2, a[1] * 2
        x2, y2 = a[2] * 2, a[3] * 2

        for x in range(x1, x2 + 1):
            for y in range(y1, y2 + 1):
                if x == x1 or x == x2 or y == y1 or y == y2:
                    if map[x][y] == 0:
                        map[x][y] = 1
                else:
                    map[x][y] = -1

    return bfs(characterX, characterY, itemX, itemY, N, map)
